examples/llms/RAG/retriever-evaluation-tutorial.ipynb

```diff
@@ -0,0 +1,1982 @@
+     "title": ""
+    }
+   },
+   "source": [
+    "# Retriever Evaluation Tutorial\n",
+    "\n",
+    "In mlflow 2.8.0, we introduced a new model type \"retriever\" to the `mlflow.evaluate()` API. It helps you to evaluate the retriever in a RAG application. It contains two built-in metrics `precision_at_k` and `recall_at_k`. As of Nov 22, 2023, in mlflow master, `ndcg_at_k` is available.\n",
+    "\n",
+    "This notebook illustrates how to use `mlflow.evaluate()` to evaluate the retriever in a RAG application. It has the following sections:\n",
+    "\n",
+    "1. Evaluation Dataset Preparation\n",
+    "1. Calling `mlflow.evaluate()`\n",
+    "1. Result Analysis and Visualization"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {
+      "byteLimit": 2048000,
+      "rowLimit": 10000
+     },
+     "inputWidgets": {},
+     "nuid": "5bf12edb-2498-4edd-aeff-b4844451850f",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "%pip install git+https://github.com/mlflow/mlflow.git langchain==0.0.339 openai faiss-cpu gensim nltk pyLDAvis tiktoken"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {
+      "byteLimit": 2048000,
+      "rowLimit": 10000
+     },
+     "inputWidgets": {},
+     "nuid": "414eb948-7f7a-411b-8308-facadb0bdde8",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [
+    {
+     "output_type": "execute_result",
+     "data": {
+      "text/plain": [
+       "'2.8.2.dev0'"
+      ]
+     },
+     "execution_count": 1,
+     "metadata": {}
+    }
+   ],
+   "source": [
+    "import ast\n",
+    "import os\n",
+    "import pprint\n",
+    "from typing import List\n",
+    "\n",
+    "import pandas as pd\n",
+    "from langchain.docstore.document import Document\n",
+    "from langchain.embeddings.openai import OpenAIEmbeddings\n",
+    "from langchain.text_splitter import CharacterTextSplitter\n",
+    "from langchain.vectorstores import FAISS\n",
+    "\n",
+    "import mlflow\n",
+    "\n",
+    "os.environ[\"OPENAI_API_KEY\"] = \"<redacted>\"\n",
+    "\n",
+    "CHUNK_SIZE = 1000\n",
+    "# Assume running from mlflow/examples/llms/RAG/\n",
+    "OUTPUT_DF_PATH = \"question_answer_source.csv\"\n",
+    "SCRAPPED_DOCS_PATH = \"mlflow_docs_scraped.csv\"\n",
+    "EVALUATION_DATASET_PATH = \"static_evaluation_dataset.csv\"\n",
+    "DB_PERSIST_DIR = \"faiss_index\"\n",
+    "\n",
+    "mlflow.__version__"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "eebcf0d9-6634-47d9-808d-e79c5a50fbbf",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "## 1. Evaluation Dataset Preparation\n",
```

Comment:
Instead of using numbering, can we just use the markdown header sections that correspond to standard html header tags? 
`##` at the top of the notebook only
`###` section header (an explanation cell)
`####` sub section within an explanation cell
`#####` as far as you probably want to go

bulleted lists or numbered lists can be auto-generated as part of nbsphinx to generate the corresponding correct html tags from within any header element.

Reply:
It's actually directly integrated into Jupyter notebooks for markdown cells. 

unordered (bullet list) 


- to get bullets
- use a hyphen

1. to get numbered lists
2. just use integers

But for the h2, h3, h4 headers, having enumerated lists look a bit odd (none of our other notebook examples use that). I think that just removing the numbers from the header sections will make the notebook tutorial feel more like the other ones.

Reply:
Got it. SGTM!